[FLOC-2278] A more reliable implementation of compute_instance_id for OpenStack

wallrj opened the pull request:
Fixes: https://clusterhq.atlassian.net/browse/FLOC-2278

adamtheturtle commented on flocker/node/agents/cinder.py:
```diff
@@ -326,21 +326,26 @@ def allocation_unit(self):
-        # and remote IPs.
+        if len(matching_instances) == 1:
+            return matching_instances[0]
+        # If there was no match, or if multiple matches were found, log an
```
`COMPUTE_INSTANCE_ID_NOT_FOUND` doesn't match with "if multiple matches were found" (though I don't know this part of the code).

jongiddy replied:
If it is possible to have no active servers (no match) *and* it is possible to have one active and one inactive server found, is it therefore possible to have no active servers and one inactive server found?  This would not get caught as an error here.

wallrj replied:
I know it's not the ideal name, but I'll leave it unchanged for now.

exarkun commented:
Thanks.

We need to figure out a way to test this logic.  At best right now it will be accidentally tested against a random set of cases based on the state of our OpenStack deployments at the time other tests are run. :/

I wonder if that OpenStack mocking library would help at all here or if it doesn't cover the APIs we need (though we could expand it).

Please file a follow-up issue for adding unit test coverage for this method and then, if the tests pass on the relevant builders, merge.

exarkun commented on flocker/node/agents/cinder.py:
```diff
@@ -326,21 +326,26 @@ def allocation_unit(self):
+            return matching_instances[0]
+        # If there was no match, or if multiple matches were found, log an
+        # error containing all the local and remote IPs.
         COMPUTE_INSTANCE_ID_NOT_FOUND(local_ips=local_ips, api_ips=api_ip_map)
```
missing `.write()` :cry: :cry: :cry:

wallrj replied:
Fixed!

wallrj commented:
Ok. I ran the API acceptance tests manually after addressing the review comments.

```
(2278)(release-maintenance/flocker-1.0.0pre1/reliable-instance-id-FLOC-2278 ? =)[~/.../HybridLogic/flocker]$ ./admin/run-acceptance-tests --no-pull --keep --distribution=centos-7 --provider=managed --config=acceptance.yml  --dataset-backend=rackspace --branch=release-maintenance/flocker-1.0.0pre1/reliable-instance-id-FLOC-2278 --flocker-version='' flocker.acceptance.test_api

...
flocker.acceptance.test_api
  ContainerAPITests
    test_create_container_with_dataset ...                                 [OK]
    test_create_container_with_environment ...                          [ERROR]
                         [ERROR]
    test_create_container_with_ports ...                                   [OK]
    test_current ...                                                       [OK]
    test_linking ...                                                       [OK]
    test_move_container_with_dataset ...                                   [OK]
    test_non_root_container_can_access_dataset ...                         [OK]
  DatasetAPITests
    test_dataset_creation ...                                              [OK]
    test_dataset_deletion ...                                              [OK]
    test_dataset_move ...                                                  [OK]

===============================================================================
[ERROR]
Traceback (most recent call last):
Failure: twisted.internet.defer.TimeoutError: <flocker.acceptance.test_api.ContainerAPITests testMethod=test_create_container_with_environment> (test_create_container_with_environment) still running at 120.0 secs

flocker.acceptance.test_api.ContainerAPITests.test_create_container_with_environment
===============================================================================
[ERROR]
Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
DelayedCalls: (set twisted.internet.base.DelayedCall.debug = True to debug)
<DelayedCall 0x2f5b518 [0.0061137676239s] called=0 cancelled=0 Deferred.callback(None)>

flocker.acceptance.test_api.ContainerAPITests.test_create_container_with_environment
-------------------------------------------------------------------------------
Ran 10 tests in 729.585s

FAILED (errors=2, successes=9)
--keep specified, not destroying nodes.
To run acceptance tests against these nodes, set the following environment variables: 
export FLOCKER_ACCEPTANCE_VOLUME_BACKEND=openstack;
export FLOCKER_ACCEPTANCE_NUM_AGENT_NODES=2;
export FLOCKER_ACCEPTANCE_HOSTNAME_TO_PUBLIC_ADDRESS='{}';
export FLOCKER_ACCEPTANCE_API_CERTIFICATES_PATH=/tmp/tmpcfbot8;
export FLOCKER_ACCEPTANCE_CONTROL_NODE=119.9.105.67;
(2278)(release-maintenance/flocker-1.0.0pre1/reliable-instance-id-FLOC-2278 ? =)[~/.../HybridLogic/flocker]$ 

```

They all passed, apart from one spurious error.

So I'll merge this.
